src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -1456,487 +114,88 @@ internal static unsafe char ParseFormatSpecifier(ReadOnlySpan<char> format, out
-
-                                    while (++src < format.Length && pFormat[src] == '0')
-                                        i++;
-                                    if (i > 10)
-                                        i = 10;
-
-                                    int exp = dig[0] == 0 ? 0 : number.scale - decimalPos;
-                                    FormatExponent(ref sb, info, exp, ch, i, positiveSign);
-                                    scientific = false;
-                                }
-                                else
-                                {
-                                    sb.Append(ch); // Copy E or e to output
-                                    if (src < format.Length)
-                                    {
-                                        if (pFormat[src] == '+' || pFormat[src] == '-')
-                                            sb.Append(pFormat[src++]);
-                                        while (src < format.Length && pFormat[src] == '0')
-                                            sb.Append(pFormat[src++]);
-                                    }
-                                }
+                        case ';':
+                            if (--section != 0)
                                 break;
-                            }
-                        default:
-                            sb.Append(ch);
-                            break;
+                            if (src < format.Length && pFormat[src] != 0 && pFormat[src] != ';')
+                                return src;
+                            goto case '\0';
+                        case '\0':
+                            return 0;
                     }
                 }
             }
+        }
+
+        private static unsafe void FormatExponent(ref ValueStringBuilder sb, NumberFormatInfo info, int value, char expChar, int minDigits, bool positiveSign)
+        {
+            sb.Append(expChar);
+
+            if (value < 0)
+            {
+                sb.Append(info.NegativeSign);
+                value = -value;
+            }
+            else
+            {
+                if (positiveSign)
+                    sb.Append(info.PositiveSign);
+            }
 
-            if (number.sign && (section == 0) && (number.scale == 0) && (sb.Length > 0))
-                sb.Insert(0, info.NegativeSign);
+            char* digits = stackalloc char[MaxUInt32DecDigits];
+            char* p = UInt32ToDecChars(digits + MaxUInt32DecDigits, (uint)value, minDigits);
+            int i = (int)(digits + MaxUInt32DecDigits - p);
+            sb.Append(p, (int)(digits + MaxUInt32DecDigits - p));
         }
 
-        private static void FormatCurrency(ref ValueStringBuilder sb, ref NumberBuffer number, int nMaxDigits, NumberFormatInfo info)
+        private static unsafe void FormatCurrency(ref ValueStringBuilder sb, bool sign, int scale, char* dig, int nMaxDigits, NumberFormatInfo info)
```

Comment:
The `Format*` functions previously took a `ref NumberBuffer`; they were changed to take the `sign`, `scale`, and `digit pointer` individually so they could be shared between the `IntegerBuffer` and `NumberBuffer` implementations.